hazelcast-client/src/test/java/com/hazelcast/client/heartbeat/ClientHeartbeatTest.java

```diff
@@ -211,4 +218,66 @@ private static ClientConfig getClientConfig() {
         clientConfig.setProperty(ClientProperty.HEARTBEAT_INTERVAL.getName(), "500");
         return clientConfig;
     }
+
+
+    @Test
+    public void testAuthentication_whenHeartbeatResumed() throws Exception {
+        HazelcastInstance hazelcastInstance = hazelcastFactory.newHazelcastInstance();
+        ClientConfig config = new ClientConfig();
+        config.setProperty(ClientProperty.SHUFFLE_MEMBER_LIST.getName(), "false");
+        final HazelcastInstance client = hazelcastFactory.newHazelcastClient(config);
+        HazelcastClientInstanceImpl hazelcastClientInstanceImpl = getHazelcastClientInstanceImpl(client);
+        final ClusterListenerSupport clientClusterService =
+                (ClusterListenerSupport) hazelcastClientInstanceImpl.getClientClusterService();
+
+        final CountDownLatch countDownLatch = new CountDownLatch(2);
+        client.getLifecycleService().addLifecycleListener(new LifecycleListener() {
+            @Override
+            public void stateChanged(LifecycleEvent event) {
+                countDownLatch.countDown();
+            }
+        });
+
+        HazelcastInstance instance2 = hazelcastFactory.newHazelcastInstance();
+        blockMessagesFromInstance(instance2, client);
+
+        final HazelcastInstance instance3 = hazelcastFactory.newHazelcastInstance();
+        hazelcastInstance.shutdown();
+
+        //wait for disconnect from instance1 since it is shutdown  // CLIENT_DISCONNECTED event
+        //and wait for connect to from instance3 // CLIENT_CONNECTED event
+        assertOpenEventually(countDownLatch);
+
+        //verify and wait for authentication to 3 is complete
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                assertEquals(instance3.getLocalEndpoint().getUuid(), clientClusterService.getPrincipal().getOwnerUuid());
+                assertEquals(instance3.getCluster().getLocalMember().getAddress(), clientClusterService.getOwnerConnectionAddress());
+            }
+        }, 10);
+
+        //unblock instance 2 for authentication response.
+        unblockMessagesFromInstance(instance2, client);
+
+        //trigger incoming messages from instance2
+        IExecutorService executorService = client.getExecutorService("s");
+        Future f = executorService.submitToMember(new SerializableClass(), instance2.getCluster().getLocalMember());
+        f.get();
+
+        //late authentication response from instance2 should not be able to change clients state
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                assertEquals(instance3.getLocalEndpoint().getUuid(), clientClusterService.getPrincipal().getOwnerUuid());
```

Comment:
can you check that the owner membership map has the correct owner at both instance2 and instance3 for the the client.